Do task

YakovMalykh opened the pull request.

YakovMalykh committed:
прописал ДБ и ключ бота в application.properties

YakovMalykh committed:
выполнил основную задачу

YakovMalykh committed:
закончил логику, которую хотел и написал тесты... протестировать удалось не все методы, есть сложность
с тестами методов использующих класс Update(см комментарии в ConstantsTestBot

OlgaLugacheva commented on telegram-bot/src/main/java/pro/sky/telegrambot/interfacies/NotificationService.java:
```diff
@@ -0,0 +1,30 @@
+public interface NotificationService {
+    SendResponse greeting(Update update);
+
+    public Notifications saveNotificationToDB(Update update);
+    public List<Notifications> checkCurrentNotifications();
```
в интерфейсе и  так все методы public, поэтому принудительное указание тут избыточно

YakovMalykh replied:
исправил

OlgaLugacheva commented on telegram-bot/src/main/java/pro/sky/telegrambot/listener/TelegramBotUpdatesListener.java:
```diff
@@ -28,9 +39,80 @@ public void init() {
-            // Process your updates here
+            // проверяю на соответсвие текста в отправленном юзером сообщении
+            // требуемому сообщению
+            if (update.message().text().equals("/start")) {
```
здесь напрашивается switch case

YakovMalykh replied:
исправил

OlgaLugacheva commented on telegram-bot/src/test/java/pro/sky/telegrambot/СonstantsTestBot.java:
```diff
@@ -0,0 +1,60 @@
+
+    // взял из логов строку Update, хочу преобразовать обратно в JSON и уже из этого JSON воссоздать экземпляр класса Update
+
+//    public static final String str = "Update{update_id=900654397, message=Message{message_id=357, from=User{id=584199371, is_bot=false, first_name='Yakov', last_name='Malykh', username='null'" +
```
из произвольной строки мы не можем сформировать json, она должна соответствовать формату json-a

YakovMalykh replied:
вышел из ситуации иначе

OlgaLugacheva commented on telegram-bot/src/test/java/pro/sky/telegrambot/NotificationServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+
+//    @Test// НЕ ЗАВЕРШЕН, нужно дописать логику
+//    public void shouldReturnGreeting() {
+//        when(updateMock.message().chat().id()).thenReturn(1L);
```
если этот тест раскомментировать, то по сути он тестирует само апи, что оно корректно работает

OlgaLugacheva commented on telegram-bot/src/test/java/pro/sky/telegrambot/NotificationServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+//    }
+
+    //        @Test
+//    public void shouldReturnNameOfUser() {
```
в getNameOfUser в качестве параметров можно сразу отдавать строки, если хочется протестировать эту логику

YakovMalykh replied:
не совсемпонял, что ты имеешь ввиду... я в параметры метода public String getNameOfUser(Update update) подаю тот же класс Update и из него достаю имя... опять же, пытаюсь его замокать сталкиваюсь с той же проблеммой 
Cannot invoke "com.pengrad.telegrambot.model.Message.from()" because the return value of "com.pengrad.telegrambot.model.Update.message()" is null

OlgaLugacheva commented on telegram-bot/src/test/java/pro/sky/telegrambot/NotificationServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+//    public void shouldReturnSavedNotification() {
+//        when(out.parsMessage(any(Update.class))).thenReturn(NOTIFICATION_4_FUTURE);
+//        when(out.notificationIsUnique(NOTIFICATION_4_FUTURE)).thenReturn(true);
+//        Notifications result = out.saveNotificationToDB(updateMock);
```
здесь можно проверить, что возвращает репозиторий

YakovMalykh replied:
здесь я тоже упираюсь в ту же проблему с Update
Cannot invoke "com.pengrad.telegrambot.model.Message.text()" because the return value of "com.pengrad.telegrambot.model.Update.message()" is null

т.е. без экземплряа класса Update в константах, я не могу протестировать методы использующие в своей логике этот класс...

OlgaLugacheva commented on telegram-bot/src/test/java/pro/sky/telegrambot/NotificationServiceImplTest.java:
```diff
@@ -0,0 +1,104 @@
+//        when(MESSAGE.chat().id()).thenReturn(1L);
+//        when(out.getNameOfUser(updateMock)).thenReturn("User 1");
+//        when(MESSAGE.from().firstName()).thenReturn("User 1");
+//        Notifications result = out.parsMessage(updateMock);
```
для тестирования парсинга проще всего вынести саму логику парсинга в отдельный метод, в который будет передаваться строка

OlgaLugacheva commented on telegram-bot/src/main/java/pro/sky/telegrambot/model/Notifications.java:
```diff
@@ -0,0 +1,101 @@
+    @Id
+//    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "chat_id")
```
JoinColumn ? зачем, если у нас тут вообще одна таблица ? если нужно указать имя колонки, то `@Column`

YakovMalykh replied:
исправил

OlgaLugacheva commented on telegram-bot/src/main/java/pro/sky/telegrambot/listener/TelegramBotUpdatesListener.java:
```diff
@@ -28,9 +39,80 @@ public void init() {
+    @Scheduled(cron = "0 00 23 * * *")
+    // метод удаляет устаревшие напоминания из БД
+    // не доработан, после чистки БД почему-то все равно выкидывает ошибку
+    // JpaSystemException: could not extract ResultSet;
```
возможно , проблема c JoinColumn , сам метод верен

YakovMalykh replied:
метода рабочий, это я забыл удалить устаревший комментарий. Теперь удалил

YakovMalykh committed:
исправил замечания, дописал проверку всех методов.
создал несколько сових классов (ConstantMyUser, ConstantMyChat, ConstantMyMessage и ConstantMyUpdate), которые наследуются от User, Char, Message и Update соответственно, чтобы создать сущность с необходимыми полями для тестов, т.к. Update замокать не удалось
